### ISSUE TITLE
Random reports of error 7022 (implicitly has type 'any' because self-reference)

### ISSUE THREAD

**Issue by voliva** [microsoft/TypeScript#51434]
# Bug Report

### 🔎 Search Terms

implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer (7022)

### 🕗 Version & Regression Information

```ts
interface MyType {
  method(mapFn: (paramA: () => string, paramB: string) => any): void;
}

const instance: MyType = null as any;

instance.method((getValue, key) => {
  // Only happens if `key` is declared on the line above
  const value = getValue();

  /**
   * If you select `.length` and type `.`, the `const value` line above will show error.
   * If you just select `length`, deleted, and ask for suggestions, everything works just fine.
   * It also works fine if you "restart TS server" (or Ctrl+S and refreshing the playground) after having the error showing up.
   */
  return value.length;
});
```

### 🙁 Actual behavior

[Context, skip if you are in a hurry]
On a big project I'm working on, sometimes happens that as I'm developing some variables would start having `unknown` values, and some others would start showing the 7022 error. The solution I always had is to restart the TS server and most of the time that solves it.

Recently I've found a consistent way of triggering this error on something that feels like it's not self-referencing in any way, which I reduced to this snippet. Before posting this I did a search and found many other bugs opened with code that works with loops, arrays, etc. that in one way or another have some self-referencing, but this one I feel it's different.

[TL;DR of this section]
On some cases TS Server emits 7022 while changing the contents of a file, in cases where there's no self-referencing variables.

On this snippet, the error doesn't show up if you remove the second argument of the function (`key`), or if you select `length`, delete it and ask for suggestions (TS server still understands it's a string and shows a list of its properties).

However, if you select `.length` and press `.`, then it emits 7022 and doesn't know what's the type of `value` (eventhough if you hover over `getValue()` to see its type, you can see it's just `() => string`).

If you restart the TS server at this point (or the playground by saving + reloading), it will show a parse error (because the code is not finished), but the type of `value` will correctly be `string`.

(Note that in my big project, sometimes TS would still show up 7022 errors even when the file was syntactically correct. Only forcing a TS Server reset would fix it)

### 🙂 Expected behavior

TS Server should not emit 7022 when there are no self-referencing variables.

**Comment by Compositr** [comment]
This issue also happens to me:
![image](https://user-images.githubusercontent.com/43405050/206704592-68f6df16-26ca-42e0-9f29-8ad4fd71637a.png)

Can confirm a TS Server restart solves it temporarily

**Comment by helmturner** [comment]
This basically breaks intellisense. 

Note that in the first image, everything is fine, and the types for `formattedOrgData` are fine.

<img width="885" alt="Screenshot 2023-01-08 at 21 22 00" src="https://user-images.githubusercontent.com/66281141/211236762-7e1087f2-b6c1-405b-82ab-dbbd1d6b8ad2.png">


BUT, when you type a single `.` at the end of the object to inspect the properties, everything breaks, and I get 7022 errors on random variables (namely: `org` and `ref`, highlighted below).


<img width="807" alt="Screenshot 2023-01-08 at 21 21 11" src="https://user-images.githubusercontent.com/66281141/211236768-c505e0d9-cce4-41f4-a983-9b0b7314c898.png">


I've seen pretty dismissive responses from @RyanCavanaugh for other issues regarding error 7022 false positives with no clear circularity... but this is clearly a regression.

**Comment by Compositr** [comment]
Yeah, I've noticed it sometimes breaks when typing a dot on larger projects for me. I agree that this kind of nullifies the point of IntelliSense if it can't help you the **instant you need it**

**Comment by voliva** [comment]
It's been a while since I haven't seen this error in my project, and I just checked with the playground I've provided and it looks like I can't reproduced on version 5.0.4 onward, so I would say it's fixed.

Is it worth keeping this issue open so that someone might add investigate and add a test so that it doesn't regress?

**Comment by tobz1000** [comment]
I can still reproduce this with TS 5.3.2:

```ts
let a: number | undefined = undefined;

while (true) {
  // Error ts(7022) shown on `b`
  const b = a;
  a = b;
}
```

Replacing the first line with `let a: number | undefined` (no explicit assignment) does not cause the error.

**Comment by Squidgical** [comment]
Another case of this issue;

```ts
declare function func<Names extends string>(
  obj: Record<Names, () => string>,
): Record<Names, string>

declare function irrelevantArg(arg: string): string

const a = func({
  //    ^
  // 'a' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.deno-ts(7022)
  /* no it isn't; to define `a` does not require calling `rule` on the arg to `b` */
  rule: () => irrelevantArg(b.rule),
//^^^^
// 'rule' implicitly has return type 'any' because it does not have a return type annotation and is referenced directly or indirectly in one of its return expressions.deno-ts(7023)
/* no it isn't; `b.rule` has no bearing on the return type of `irrelevantArg` and 
therefore no bearing on the return type of `rule`, which is always `string` */
})
const b = func({
  rule: () => irrelevantArg(a.rule),
})
```

The complaint typescript has suggests that the type of `a` and `b` are dependent on eachother, however the types of `a` and `b` are entirely defined by the object keys, which do not reference anything external.

Some research suggests that for this particular case, the cause is that typescript won't validate my keys and make them available until it has validated the returntype of the function in the value of those keys, and that it won't validate those return types until it has validated the type of everything in the return expression. To reiterate; that is a sub-expression which doesn't contribute to the type impacting the availability of keys on the object which it's function is defined in. Like an infection spreading just by proximity, not interaction.
